### DISCUSSION TITLE
pyray get static unsigned int array[] from raylib

### DISCUSSION THREAD

**Discussion by Axle-Ozz-i-sofT** [electronstudio/raylib-python-cffi#98] [General]
Hi I am trying to convert .\raygui\examples\controls_test_suite.c to pyray.

I am stuck on Line 143: `ClearBackground(GetColor(GuiGetStyle(DEFAULT, BACKGROUND_COLOR)));`
I am always getting a negative integer return and can't seam to get the correct method using ffi to return it from the library as an unsigned integer.
The following is a test extract that I am using to try to debug and work it out:
```
from pyray import *

def la_dbg_pause():
    print("DEBUG Pause! Press [Enter] to continue...")
    input()# Press [Enter] to continue.
    return None

screenWidth = 800
screenHeight = 450

gui_style = ffi.new("int *", 1536)  # Tried several formats here without success.

init_window(screenWidth, screenHeight, "raygui - controls test suite")

print(DEFAULT)
print(BACKGROUND_COLOR)

gui_style = gui_get_style(DEFAULT, BACKGROUND_COLOR)
print(gui_style)  #DEBUG - returns negative value

#print(gui_get_style(DEFAULT, BACKGROUND_COLOR))  #DEBUG - returns negative value

la_dbg_pause()

clear_background(get_color(gui_get_style(DEFAULT, BACKGROUND_COLOR)))  # From examples/controls_test_suite.c (converted)
    
close_window()        # Close window and OpenGL context

## From raylib.h
## static unsigned int guiStyle[RAYGUI_MAX_CONTROLS*(RAYGUI_MAX_PROPS_BASE + RAYGUI_MAX_PROPS_EXTENDED)] = { 0 };

#// Get control style property value
#int GuiGetStyle(int control, int property)
#{
#    if (!guiStyleLoaded) GuiLoadStyleDefault();
#    return guiStyle[control*(RAYGUI_MAX_PROPS_BASE + RAYGUI_MAX_PROPS_EXTENDED) + property];
#}
```
Output:
```
0
19
-168430081
DEBUG Pause! Press [Enter] to continue...
```
Any small example on retrieving the value of that array would be helpfull :)
Axle

**Comment by sDos280** [comment]
firstly, please try not to open discussions about these kinds of things, open an issue instead...
secondly, when you are calling for help here try to explain better what you are trying to do(because this way I can try to get you better help)...
thirdly, using the ffi module, in that case, can be a bit pointless(+ I don't know how to work with him, try to use the ctypes for this kind of thing), also using raylib(cffi) in that case will be much easier than using pyray.
here is my attempt to make this code work:
```python
from raylib import *
from ctypes import *

def la_dbg_pause():
    print("DEBUG Pause! Press [Enter] to continue...")
    input()  # Press [Enter] to continue.
    return None


screenWidth = 800
screenHeight = 450

gui_style = c_uint(1564)

InitWindow(screenWidth, screenHeight, b"raygui - controls test suite")

print(DEFAULT)
print(BACKGROUND_COLOR)
gui_style = c_uint(GuiGetStyle(DEFAULT, BACKGROUND_COLOR))
print(gui_style)  # DEBUG - returns negative value

# print(gui_get_style(DEFAULT, BACKGROUND_COLOR))  #DEBUG - returns negative value

#la_dbg_pause()

ClearBackground(GetColor(gui_style.value))  # From examples/controls_test_suite.c (converted)

CloseWindow()  # Close window and OpenGL context

## From raylib.h
## static unsigned int guiStyle[RAYGUI_MAX_CONTROLS*(RAYGUI_MAX_PROPS_BASE + RAYGUI_MAX_PROPS_EXTENDED)] = { 0 };

# // Get control style property value
# int GuiGetStyle(int control, int property)
# {
#    if (!guiStyleLoaded) GuiLoadStyleDefault();
#    return guiStyle[control*(RAYGUI_MAX_PROPS_BASE + RAYGUI_MAX_PROPS_EXTENDED) + property];
# }
```

**Comment by sDos280** [comment]
in python, we don't really have arrays, Nevertheless using ctypes you can create an array(in the example I will present an unsigned int array) object by using ctypes
```python
from ctypes import *
number_of_elements_in_array = 5
my_array = (c_uint * number_of_elements_in_array )()
```

**Comment by sDos280** [comment]
static vairables in c are just globales vairables in python(i **think**).

**Comment by electronstudio** [comment]
> ## static unsigned int guiStyle[RAYGUI_MAX_CONTROLS*(RAYGUI_MAX_PROPS_BASE + RAYGUI_MAX_PROPS_EXTENDED)] = { 0 };
> 
> #int GuiGetStyle(int control, int property)

Basically the problem is the signatures dont match.  It's an unsigned int, but then in the second signature it's a signed int.  So CFFI trys to convert it.  You could report this as a bug, but they would probably say "RayGUI is just a simple experiment for my personal projects, you are advised to use ImGUI instead."

**Reply by electronstudio** [reply]
You could try adding +2**32 to gui_styl to convert it yourself.

**Reply by sDos280** [reply]
> You could try adding +2**32 to gui_styl to convert it yourself.

adding 2**32 to the number won't make the performance go down by a lot?

**Reply by electronstudio** [reply]
No, it won't.  But if your project requires high performance you shouldn't be using Python anyway.

**Reply by sDos280** [reply]
yeah... 🙃

**Reply by Axle-Ozz-i-sofT** [reply]
Thanks peoples
Turned out to be the 2's compliment converting C uint to Py int. I traced it through in C... (n+ 2**32) or (n +1<<32)
```
printf("%u", GuiGetStyle(DEFAULT, BACKGROUND_COLOR));
Output:4126537215
printf("%x", GuiGetStyle(DEFAULT, BACKGROUND_COLOR));
f5f5f5ff
```
It expects decimal value of the hex so...
`clear_background(get_color(2**32 + (gui_get_style(DEFAULT, BACKGROUND_COLOR))))`
is correct
```
#gui_style = 2**32 +(gui_get_style(DEFAULT, BACKGROUND_COLOR))
#print(hex(gui_style))
Output:0xf5f5f5ff
```
Just worked it out an hour back before coming to check here :) @sDos280 was on to it :P [edit] oops. meant @electronstudio

I am attempting to get C from the example:
`const char *listViewExList[8] = { "This", "is", "a", "list view", "with", "disable", "elements", "amazing!" };`
to
`int_list_struct_view_ex_active = gui_list_view_ex(Rectangle( 165, 180, 140, 200 ), list_list_view_ex_list, 8, int_list_view_ex_focus, int_list_view_ex_scroll_index, int_list_view_ex_active)  #list_list_view_ex_list`

I have tried numerous incantations from cffi.readthedocs.io but can't find an example of constructing the 2D char array :(

Sorry if I am posting in the wrong place. It didn't seam like a bug. Some weakness by me learning raylib, pyray and cffi ++ all at the same time lol

**Comment by sDos280** [comment]
using ctypes you can create a 2d array like this:
```
2d_array = (columns_num * ctypes_type) * rows
```
if you use that note that you will still need to fill the array data by yourself(manually)

**Reply by sDos280** [reply]
as I saw it, cffi uses the ctypes types with it, so using ctypes with pyray should be fine...

**Reply by Axle-Ozz-i-sofT** [reply]
Thanks @sDos280
I'll run with that (2d_array = (columns_num * ctypes_type) * rows) for the moment as a test.

cffi does have the ability to create multidimensional char arrays, but I think I am going about it the wrong way. There is a 2D **int** array example, but I think I have to populate the array after defining it the way you have suggested with ctypes above.

aka, I have been attempting to treat it as a static array rather than a dynamic memory allocation..

**Reply by Axle-Ozz-i-sofT** [reply]
Any complete examples on correctly creating an array of strings with either cffi or ctypes?

I have tried every example off every forum I can find and nothing allows me to create and populate the array :(

**Reply by sDos280** [reply]
> as I saw it, cffi uses the ctypes types with it, so using ctypes with pyray should be fine...

ok, but before you try that i made a mistake you will need to run
```python
_2d_array = ((columns_num * ctypes_type) * rows)()
```

p.s. I forgot to send that, so I send that now

**Reply by sDos280** [reply]
> Any complete examples on correctly creating an array of strings with either cffi or ctypes?
> 
> I have tried every example off every forum I can find and nothing allows me to create and populate the array :(

you need to understand that c doesn't really have strings(like python has), so when you want to use strings in c you will use char pointers.
 here is an example of creating a char array using ctypes(I think this should work)
 ```python
 array_size = 5
 charPtr_array_obj = array_size * c_char_p
 your_charPtr_array_name = charPtr_array_obj(b"here", b"is", b"the", b"example", b"!!!")
 ```
 or you can use build-in string objects in ctypes(but I never used that so I don't have an example for that)

**Reply by Axle-Ozz-i-sofT** [reply]
Thanks @sDos280
Yeah, I am predominantly from a C background so Py is still a weakness for me atm. Strings in C really means a "string of char" {'S','t','r','i','n','g','\0'} or literal char * MyString = "String".

The method you gave above to populate the array may be what I was looking for. I'll give that method a try and have a few more shots with populating `_2d_array = ((columns_num * ctypes_type) * rows)()`; row should be number of elements and col should be length of string of char (rowise I think). I seen some examples (most) using  `myarray = _2d_array` and then myarray[n].value = 'n element" but that didn't work or make any sense in context.

I appreciate the suggestions and help. I get these examples from raysan/raylib converted/working I'll gift them to raylib_cffi for others to use.

**Reply by Axle-Ozz-i-sofT** [reply]
P.S. Appears solved with the following:
```
list1 = ["This", "is", "a", "list view", "with", "disable", "elements", "amazing!"]
items = [ffi.new("char[]", x.encode('utf-8')) for x in list1]
list_list_view_ex_list = ffi.new("char *[]", items)
```
So like we would for dynamic memory allocation in C we have to loop through each dimension of the array as we create it. Aka create a list of "string of char" then populate a new ctype/cffi char* from the list giving us array[][].
I have to check if the '\0' string terminator is being passed to the C array but looks good :)
Axle

**P.S. Success :)**
![2022-10-18 12 12 21](https://user-images.githubusercontent.com/57524738/196319394-d7686f93-d588-4460-b4b8-9d433d9577af.png)

[edit]
+ source if anyone wants to make use of it. Some icons don't appear to display correctly atm
[controls_test_suite.py.txt](https://github.com/electronstudio/raylib-python-cffi/files/9806286/controls_test_suite.py.txt)

**Reply by electronstudio** [reply]
thanks, i have used your code to automate it https://github.com/electronstudio/raylib-python-cffi/commit/81590eb8158fdff3bfa540ff3c9c995cb45b414a

**Reply by Axle-Ozz-i-sofT** [reply]
Thanks, I will review my example and test it :)

**Comment by Axle-Ozz-i-sofT** [comment]
Hi a small update for /raygui/examples/controls_test_suite/controls_test_suite.py , and I'll throw in the /raylib/examples/shapes/shapes_basic_shapes.py as well.

As a last item, is it OK if translate the c to py examples using the type prefix for variables? else I will leave the prefix out for copies I give to you.

I'll likely convert as many examples as I can as a "raylib API familiarity exercise", so I may as well send them your ways as a thanks for the assist. Have to start a new tertiary intake from today so may be a while, but I'll trickle along with them in the background. I have to do conversions for FreeBASIC as well because I like to punish myself with over commitment lol

With thanks
Axle
P.S. I tested the scripts in Lubuntu 20.04LTS as well as native Windows10 and they worked fine.
[controls_test_suite.py.txt](https://github.com/electronstudio/raylib-python-cffi/files/9808948/controls_test_suite.py.txt)
[shapes_basic_shapes.py.txt](https://github.com/electronstudio/raylib-python-cffi/files/9808966/shapes_basic_shapes.py.txt)

**Reply by Axle-Ozz-i-sofT** [reply]
Current full working versions with corrections available at:
[/raylib/Python_3](https://github.com/Axle-Ozz-i-sofT/A-BEGINNERS-GUIDE-TO-PROGRAMMING/tree/main/3_Libraries_Overview/Tests_source/raylib/Python_3)

Corrected some of the ffi.new(types) so all functions return the correct "By Reference" values for arguments sent to the C library functions. Corrected some type casting and round() for floats in GUI Sliders.
